### ISSUE TITLE
Problem building caffe

### ISSUE THREAD

**Issue by isloven** [negativo17/cuda-gcc#1]
It looks like cuda-gcc under Fedora 27 has issue building caffe deep learning framework which compiles fine under compat-gcc-53 from copr. I suspect that cuda-gcc + cuda-gcc-c++ needs to be rebuilt w/ -fPIC?

Here is the error during linking:

```
cuda-g++ -shared -o .build_release/lib/libcaffe.so.1.0.0 .build_release/src/caffe/proto/caffe.pb.o .build_release/src/caffe/data_transformer.o .build_release/src/caffe/internal_thread.o .build_release/src/caffe/layer.o .build_release/src/caffe/common.o .build_release/src/caffe/layers/embed_layer.o .build_release/src/caffe/layers/cudnn_softmax_layer.o .build_release/src/caffe/layers/recurrent_layer.o .build_release/src/caffe/layers/cudnn_tanh_layer.o .build_release/src/caffe/layers/contrastive_loss_layer.o .build_release/src/caffe/layers/reduction_layer.o .build_release/src/caffe/layers/batch_reindex_layer.o .build_release/src/caffe/layers/im2col_layer.o .build_release/src/caffe/layers/memory_data_layer.o .build_release/src/caffe/layers/batch_norm_layer.o .build_release/src/caffe/layers/silence_layer.o .build_release/src/caffe/layers/slice_layer.o .build_release/src/caffe/layers/flatten_layer.o .build_release/src/caffe/layers/elu_layer.o .build_release/src/caffe/layers/scale_layer.o .build_release/src/caffe/layers/multinomial_logistic_loss_layer.o .build_release/src/caffe/layers/lrn_layer.o .build_release/src/caffe/layers/sigmoid_cross_entropy_loss_layer.o .build_release/src/caffe/layers/cudnn_lrn_layer.o .build_release/src/caffe/layers/neuron_layer.o .build_release/src/caffe/layers/data_layer.o .build_release/src/caffe/layers/deconv_layer.o .build_release/src/caffe/layers/base_data_layer.o .build_release/src/caffe/layers/tile_layer.o .build_release/src/caffe/layers/power_layer.o .build_release/src/caffe/layers/lstm_unit_layer.o .build_release/src/caffe/layers/input_layer.o .build_release/src/caffe/layers/eltwise_layer.o .build_release/src/caffe/layers/hdf5_output_layer.o .build_release/src/caffe/layers/filter_layer.o .build_release/src/caffe/layers/cudnn_pooling_layer.o .build_release/src/caffe/layers/accuracy_layer.o .build_release/src/caffe/layers/cudnn_relu_layer.o .build_release/src/caffe/layers/cudnn_conv_layer.o .build_release/src/caffe/layers/cudnn_lcn_layer.o .build_release/src/caffe/layers/parameter_layer.o .build_release/src/caffe/layers/relu_layer.o .build_release/src/caffe/layers/inner_product_layer.o .build_release/src/caffe/layers/mvn_layer.o .build_release/src/caffe/layers/concat_layer.o .build_release/src/caffe/layers/hinge_loss_layer.o .build_release/src/caffe/layers/crop_layer.o .build_release/src/caffe/layers/dropout_layer.o .build_release/src/caffe/layers/conv_layer.o .build_release/src/caffe/layers/log_layer.o .build_release/src/caffe/layers/exp_layer.o .build_release/src/caffe/layers/spp_layer.o .build_release/src/caffe/layers/bias_layer.o .build_release/src/caffe/layers/image_data_layer.o .build_release/src/caffe/layers/hdf5_data_layer.o .build_release/src/caffe/layers/prelu_layer.o .build_release/src/caffe/layers/reshape_layer.o .build_release/src/caffe/layers/dummy_data_layer.o .build_release/src/caffe/layers/pooling_layer.o .build_release/src/caffe/layers/split_layer.o .build_release/src/caffe/layers/threshold_layer.o .build_release/src/caffe/layers/rnn_layer.o .build_release/src/caffe/layers/lstm_layer.o .build_release/src/caffe/layers/softmax_layer.o .build_release/src/caffe/layers/loss_layer.o .build_release/src/caffe/layers/base_conv_layer.o .build_release/src/caffe/layers/argmax_layer.o .build_release/src/caffe/layers/cudnn_sigmoid_layer.o .build_release/src/caffe/layers/euclidean_loss_layer.o .build_release/src/caffe/layers/sigmoid_layer.o .build_release/src/caffe/layers/window_data_layer.o .build_release/src/caffe/layers/tanh_layer.o .build_release/src/caffe/layers/absval_layer.o .build_release/src/caffe/layers/infogain_loss_layer.o .build_release/src/caffe/layers/softmax_loss_layer.o .build_release/src/caffe/layers/bnll_layer.o .build_release/src/caffe/syncedmem.o .build_release/src/caffe/blob.o .build_release/src/caffe/util/signal_handler.o .build_release/src/caffe/util/cudnn.o .build_release/src/caffe/util/blocking_queue.o .build_release/src/caffe/util/benchmark.o .build_release/src/caffe/util/im2col.o .build_release/src/caffe/util/db_leveldb.o .build_release/src/caffe/util/math_functions.o .build_release/src/caffe/util/db_lmdb.o .build_release/src/caffe/util/hdf5.o .build_release/src/caffe/util/upgrade_proto.o .build_release/src/caffe/util/insert_splits.o .build_release/src/caffe/util/db.o .build_release/src/caffe/util/io.o .build_release/src/caffe/parallel.o .build_release/src/caffe/layer_factory.o .build_release/src/caffe/net.o .build_release/src/caffe/solver.o .build_release/src/caffe/solvers/adagrad_solver.o .build_release/src/caffe/solvers/nesterov_solver.o .build_release/src/caffe/solvers/adadelta_solver.o .build_release/src/caffe/solvers/rmsprop_solver.o .build_release/src/caffe/solvers/adam_solver.o .build_release/src/caffe/solvers/sgd_solver.o .build_release/cuda/src/caffe/layers/dropout_layer.o .build_release/cuda/src/caffe/layers/concat_layer.o .build_release/cuda/src/caffe/layers/absval_layer.o .build_release/cuda/src/caffe/layers/bias_layer.o .build_release/cuda/src/caffe/layers/cudnn_softmax_layer.o .build_release/cuda/src/caffe/layers/batch_reindex_layer.o .build_release/cuda/src/caffe/layers/sigmoid_layer.o .build_release/cuda/src/caffe/layers/lstm_unit_layer.o .build_release/cuda/src/caffe/layers/hdf5_data_layer.o .build_release/cuda/src/caffe/layers/tile_layer.o .build_release/cuda/src/caffe/layers/embed_layer.o .build_release/cuda/src/caffe/layers/cudnn_pooling_layer.o .build_release/cuda/src/caffe/layers/accuracy_layer.o .build_release/cuda/src/caffe/layers/pooling_layer.o .build_release/cuda/src/caffe/layers/mvn_layer.o .build_release/cuda/src/caffe/layers/bnll_layer.o .build_release/cuda/src/caffe/layers/conv_layer.o .build_release/cuda/src/caffe/layers/power_layer.o .build_release/cuda/src/caffe/layers/cudnn_sigmoid_layer.o .build_release/cuda/src/caffe/layers/cudnn_lcn_layer.o .build_release/cuda/src/caffe/layers/cudnn_relu_layer.o .build_release/cuda/src/caffe/layers/softmax_loss_layer.o .build_release/cuda/src/caffe/layers/filter_layer.o .build_release/cuda/src/caffe/layers/sigmoid_cross_entropy_loss_layer.o .build_release/cuda/src/caffe/layers/base_data_layer.o .build_release/cuda/src/caffe/layers/im2col_layer.o .build_release/cuda/src/caffe/layers/hdf5_output_layer.o .build_release/cuda/src/caffe/layers/prelu_layer.o .build_release/cuda/src/caffe/layers/scale_layer.o .build_release/cuda/src/caffe/layers/cudnn_tanh_layer.o .build_release/cuda/src/caffe/layers/contrastive_loss_layer.o .build_release/cuda/src/caffe/layers/inner_product_layer.o .build_release/cuda/src/caffe/layers/softmax_layer.o .build_release/cuda/src/caffe/layers/cudnn_lrn_layer.o .build_release/cuda/src/caffe/layers/silence_layer.o .build_release/cuda/src/caffe/layers/lrn_layer.o .build_release/cuda/src/caffe/layers/reduction_layer.o .build_release/cuda/src/caffe/layers/batch_norm_layer.o .build_release/cuda/src/caffe/layers/cudnn_conv_layer.o .build_release/cuda/src/caffe/layers/elu_layer.o .build_release/cuda/src/caffe/layers/deconv_layer.o .build_release/cuda/src/caffe/layers/tanh_layer.o .build_release/cuda/src/caffe/layers/eltwise_layer.o .build_release/cuda/src/caffe/layers/split_layer.o .build_release/cuda/src/caffe/layers/crop_layer.o .build_release/cuda/src/caffe/layers/threshold_layer.o .build_release/cuda/src/caffe/layers/exp_layer.o .build_release/cuda/src/caffe/layers/recurrent_layer.o .build_release/cuda/src/caffe/layers/log_layer.o .build_release/cuda/src/caffe/layers/relu_layer.o .build_release/cuda/src/caffe/layers/slice_layer.o .build_release/cuda/src/caffe/layers/euclidean_loss_layer.o .build_release/cuda/src/caffe/util/math_functions.o .build_release/cuda/src/caffe/util/im2col.o .build_release/cuda/src/caffe/solvers/sgd_solver.o .build_release/cuda/src/caffe/solvers/rmsprop_solver.o .build_release/cuda/src/caffe/solvers/adam_solver.o .build_release/cuda/src/caffe/solvers/adadelta_solver.o .build_release/cuda/src/caffe/solvers/adagrad_solver.o .build_release/cuda/src/caffe/solvers/nesterov_solver.o -Wl,-soname,libcaffe.so.1.0.0 -Wl,-rpath,\$ORIGIN/../lib -pthread -fPIC -DCAFFE_VERSION=1.0.0 -DNDEBUG -O2 -DUSE_CUDNN -DUSE_OPENCV -DUSE_LEVELDB -DUSE_LMDB -I/usr/include/python2.7 -I/usr/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/cuda -I.build_release/src -I./src -I./include -I/usr/include -I/usr/include/openblas -Wall -Wno-sign-compare -L/usr/lib -L/usr/local/lib -L/usr/lib -L/usr/lib64 -L/usr/lib -L.build_release/lib  -lcudart -lcublas -lcurand -lglog -lgflags -lprotobuf -lboost_system -lboost_filesystem -lm -lhdf5_hl -lhdf5 -lleveldb -lsnappy -llmdb -lopencv_core -lopencv_highgui -lopencv_imgproc -lopencv_imgcodecs -lboost_thread -lstdc++ -lcudnn -lopenblas
/usr/bin/ld: /usr/lib64/gcc/x86_64-redhat-linux/6.4.0/libstdc++.a(eh_globals.o): relocation R_X86_64_TPOFF32 against `_ZZN12_GLOBAL__N_110get_globalEvE6global' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /usr/lib64/gcc/x86_64-redhat-linux/6.4.0/libstdc++.a(cow-shim_facets.o): relocation R_X86_64_PC32 against symbol `_ZTVSt7collateIcE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
```

**Comment by scaronni** [comment]
I'm making a test right now with a build with shared objects enabled and private libraries. Would you care to test it and provide some feedback? The packages are not in a repository, so:

```
rpm -Uvh \
  https://negativo17.org/repos/_temp/cuda-gcc-6.4.0-3.fc27.x86_64.rpm \
  https://negativo17.org/repos/_temp/cuda-gcc-c++-6.4.0-3.fc27.x86_64.rpm
```
SPEC files with the changes are here:

https://github.com/negativo17/cuda-gcc

**Comment by isloven** [comment]
First off, thank you for the help!

It now errors out with this:
/usr/bin/ld: cannot find -lgcc_s
collect2: error: ld returned 1 exit status

Which is a bit strange because it looks like that library is in the package:
$ rpmls cuda-gcc | grep libgcc_s.so*
-rw-r--r--  /usr/lib64/gcc/x86_64-redhat-linux/lib64/libgcc_s.so
-rw-r--r--  /usr/lib64/gcc/x86_64-redhat-linux/lib64/libgcc_s.so.1

If I add "-L/usr/lib64/gcc/x86_64-redhat-linux/lib64" to the linker flags, it actually gets further and builds a portion of the caffe framework, although some stuff seems to fail:

```
cuda-g++ .build_release/tools/train_net.o -o .build_release/tools/train_net.bin -pthread -fPIC -DCAFFE_VERSION=1.0.0 -DNDEBUG -O2 -DUSE_CUDNN -DUSE_OPENCV -DUSE_LEVELDB -DUSE_LMDB -I/usr/include/python2.7 -I/usr/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/cuda -I.build_release/src -I./src -I./include -I/usr/include -I/usr/include/openblas -Wall -Wno-sign-compare -lcaffe -L/usr/lib -L/usr/local/lib -L/usr/lib -L/usr/lib64 -L/usr/lib -L.build_release/lib  -lcudart -lcublas -lcurand -lglog -lgflags -lprotobuf -lboost_system -lboost_filesystem -lm -lhdf5_hl -lhdf5 -lleveldb -lsnappy -llmdb -lopencv_core -lopencv_highgui -lopencv_imgproc -lopencv_imgcodecs -lboost_thread -lstdc++ -lcudnn -lopenblas -Wl,-rpath,\$ORIGIN/../lib -L/usr/lib64/gcc/x86_64-redhat-linux/lib64
//usr/lib64/libtbb.so.2: undefined reference to `std::__exception_ptr::exception_ptr::exception_ptr(void*)@CXXABI_1.3.11'
//usr/lib64/libtbb.so.2: undefined reference to `__cxa_init_primary_exception@CXXABI_1.3.11'
```

Note that if I replace "cuda-g++" with just "g++" for the linking part, it actually succeeds in building everything, although it gives off bunch of warnings.

**Comment by mrceresa** [comment]
Dear all,
thanks for your help and all your efforts in supporting cuda on fedora!
I was wondering which is the status of that, can I use the new cuda-gcc revision 3 to build tensorflow with cuda 9 or should I fall back to cuda 8 + compat-gcc-53?

Thanks and regards,
Mario

**Comment by scaronni** [comment]
Hello, could you guys try the new `6.4.0-4` build I just pushed to the same folder? I was able to build a few software components with it.

```
rpm -Uvh \
  https://negativo17.org/repos/_temp/cuda-gcc-6.4.0-4.fc27.x86_64.rpm \
  https://negativo17.org/repos/_temp/cuda-gcc-c++-6.4.0-4.fc27.x86_64.rpm
```
If it works, I will push it to the repositories along with the CUDA 9.1 update.

Thanks.

**Comment by isloven** [comment]
It looks like this version fixes the -lgcc_s problem, however the tbb issue still remains:

//usr/lib64/libtbb.so.2: undefined reference to `std::__exception_ptr::exception_ptr::exception_ptr(void*)@CXXABI_1.3.11'
//usr/lib64/libtbb.so.2: undefined reference to `__cxa_init_primary_exception@CXXABI_1.3.11'
collect2: error: ld returned 1 exit status

Full output:
```
cuda-g++ .build_release/tools/train_net.o -o .build_release/tools/train_net.bin -pthread -fPIC -DCAFFE_VERSION=1.0.0 -DNDEBUG -O2 -DUSE_CUDNN -DUSE_OPENCV -DUSE_LEVELDB -DUSE_LMDB -I/usr/include/python2.7 -I/usr/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/include -I/usr/include/cuda -I.build_release/src -I./src -I./include -I/usr/include -I/usr/include/openblas -Wall -Wno-sign-compare -lcaffe -L/usr/lib -L/usr/local/lib -L/usr/lib -L/usr/lib64 -L/usr/lib -L.build_release/lib  -lcudart -lcublas -lcurand -lglog -lgflags -lprotobuf -lboost_system -lboost_filesystem -lm -lhdf5_hl -lhdf5 -lleveldb -lsnappy -llmdb -lopencv_core -lopencv_highgui -lopencv_imgproc -lopencv_imgcodecs -lboost_thread -lstdc++ -lcudnn -lopenblas \
	-Wl,-rpath,\$ORIGIN/../lib
//usr/lib64/libtbb.so.2: undefined reference to `std::__exception_ptr::exception_ptr::exception_ptr(void*)@CXXABI_1.3.11'
//usr/lib64/libtbb.so.2: undefined reference to `__cxa_init_primary_exception@CXXABI_1.3.11'
collect2: error: ld returned 1 exit status
make: *** [Makefile:626: .build_release/tools/train_net.bin] Error 1
```

**Comment by isloven** [comment]
For the reference, here is how to replicate the build issue, if that is easier:

$ git clone https://github.com/BVLC/caffe.git
$ cd caffe
$ cp Makefile.config.example Makefile.config

Here are the changes to Makefile.config:
```
$ diff Makefile.config.example Makefile.config
5c5
< # USE_CUDNN := 1
---
> USE_CUDNN := 1
21c21
< # OPENCV_VERSION := 3
---
> OPENCV_VERSION := 3
25c25
< # CUSTOM_CXX := g++
---
> CUSTOM_CXX := cuda-g++
28c28
< CUDA_DIR := /usr/local/cuda
---
> CUDA_DIR := /usr
37,38c37
< CUDA_ARCH := -gencode arch=compute_20,code=sm_20 \
< 		-gencode arch=compute_20,code=sm_21 \
---
> CUDA_ARCH := \
51c50
< BLAS := atlas
---
> BLAS := open
55c54
< # BLAS_INCLUDE := /path/to/your/blas
---
> BLAS_INCLUDE := /usr/include/openblas
95c94
< INCLUDE_DIRS := $(PYTHON_INCLUDE) /usr/local/include
---
> INCLUDE_DIRS := $(PYTHON_INCLUDE) /usr/local/include /usr/include/cuda
121c120
< Q ?= @
---
> #Q ?= @
```

You'll need few packages installed as well, mainly glog-devel, gflags-devel, openblas-devel. 

Build:
$ make

**Comment by scaronni** [comment]
So, `tbb` in Fedora 27 is compiled with the system compiler, GCC 7.1:

https://koji.fedoraproject.org/koji/rpminfo?rpmID=12187948

So I guess you can not use a different compiler than the one on your system for this, unless you also rebuild all the other dependencies required. The `cuda-gcc` pacakge is supposed to be used as a drop in replacement for `nvcc` like here:

https://github.com/negativo17/blender/blob/fedora-27/blender-2.79-cuda.patch#L23

or here:

https://github.com/negativo17/ccminer/blob/master/ccminer.spec#L73

**Comment by scaronni** [comment]
Did it compile succesfully under Fedora 26 with `compat-gcc-53`? Because there as well there is GCC 7, and you should not be able to link it with GCC 5.3. Same as here:

https://bugs.archlinux.org/task/54944

**Comment by isloven** [comment]
Yes, at one point I've successfully compiled it with compat-gcc-53 and I've been using it ever since. I was attempting to recompile it because I've updated to CUDA 9 and some stuff got broken. 

Regardless, you already fixed the original (main) problem/issue where linking was failing because cuda-gcc was not built with -fPIC support. As I mentioned above, caffe can be linked using system g++ once the code was compiled using nvcc with the updated cuda-g++. This is excellent news!

The libtbb error is a minor inconvenience and it doesn't sound like it can be fixed easily anyway so I would not worry about it. 

So as far as I am concern, push the last update out and this issue can be closed!

Thank you very much for the help!
